toolchain/check/impl_lookup.cpp

```diff
@@ -347,6 +348,109 @@ static auto LookupImplWitnessInSelfFacetValue(
   return EvalImplLookupResult::MakeNonFinal();
 }
 
+// Substitutes witnesess in place of `LookupImplWitness` queries into `.Self`,
+// when the witness is for the same interface as the one `.Self` is referring
+// to.
+//
+// This allows access to the `FacetType` and its constraints from the witness,
+// and allows `ImplWitnessAccess` instructions to be immediately resolved to a
+// more specific value when possible.
+class SubstWitnessesCallbacks : public SubstInstCallbacks {
+ public:
+  // `context` must not be null.
+  explicit SubstWitnessesCallbacks(
+      Context* context, SemIR::LocId loc_id,
+      llvm::ArrayRef<SemIR::SpecificInterface> interfaces,
+      llvm::ArrayRef<SemIR::InstId> witness_inst_ids)
+      : SubstInstCallbacks(context),
+        loc_id_(loc_id),
+        interfaces_(interfaces),
+        witness_inst_ids_(witness_inst_ids) {}
+
+  auto Subst(SemIR::InstId& inst_id) const -> bool override {
+    // `FacetType` can be concrete even when it has rewrite constraints that
+    // have a symbolic dependency on `.Self`. See use of
+    // `GetConstantValueIgnoringPeriodSelf` in eval. So in order to recurse into
+    // `FacetType` we must not return `true` when it's concrete.
+    if (context().constant_values().Get(inst_id).is_concrete() &&
+        !context().insts().Is<SemIR::FacetType>(inst_id)) {
+      return true;
+    }
```

Comment:
Is there any risk of our getting confused by nested `where`s and substituting into the wrong `.Self`? Eg, in `A where .T = B(C where .U = V)`, we should not substitute into the `.Self.(C.U)`.

I wonder if we could reduce the risk by only substituting into the LHS and RHS of rewrite constraints rather than substituting into the whole facet type, and making `Subst` always return `false` when given a `FacetType` (which would then always be introducing a nested `.Self`).

That might also be imperfect, for cases like `A where .T = (B(.Self) where .U = ())`, where maybe we need to substitute into some of the facet type but not all of it (the `B(.Self)` but not the `.U`). I'm not sure.

I'm fine with just adding a TODO for now and looking further into this later.